rfd/0181-pg-database-access-webui.md

```diff
@@ -0,0 +1,322 @@
+
+In the new tab, he's prompted with the MFA modal (the same as SSH
+sessions). After completing the authentication, he starts performing multiple
+queries using the interactive shell, but after some time, he exceeds his session
+TTL, the connection is dropped, and he is redirected to the login page. His
+database session is terminated.
+
+#### PostgreSQL interactive terminal
+
+##### Banner
+
+At the beginning of each session, we'll include an informative banner that will consist of the following:
+- Information about the interactive shell name and its version.
+- Connected PostgreSQL version (this mimics the `psql`).
+- Parameters used to connect to the instance, like the database and username.
+- Short text describing how to get the supported commands and information through the help command (more details in the next section).
+
+```shell
+Teleport PostgreSQL interactive shell (v16.2.0-dev)
+Connected to "pg" instance (16.2) as "alice" user (with roles "read-write").
+Type "help" or \? for help.
+```
+
+##### Supported commands
+
+In addition to executing queries, the interactive shell will implement some
+backlash (\\) commands from `psql`. Users can fetch the list of supported
+commands by calling the help command (`\?` or `help`):
+
+```shell
+postgres=> \?
+General:
+  \q          Terminate the session.
+  \teleport   Show Teleport interactive shell information, such as execution limitations.
+
+Informational:
+  \d              List tables, views, and sequences.
+  \d NAME         Describe table, view, sequence, or index.
+  \dt [PATTERN]   List tables.
+
+Connection/Session:
+  \session   Display information about the current session, like user, roles, and database instance.
+```
+
+##### Limitations/Unpported commands
+
+Teleport PostgreSQL interactive shell will not be a complete feature pair with
+`psql`. Those limitations will be due to security measures or the shell's
+simplicity. Given those limitations, messages will be shown to the users,
+displaying a description and direction on executing the desired command
+(if applicable).
+
+Unsupported backslash (\\) commands from psql will only display a failure
+message:
+
+```shell
+postgres=> \set a "hello"
+Invalid command \set.
+Try "help" or "\?" for the list of supported commands.
+
+postgres=>
+```
+
+Other limitations, such as query size limit, will display a more complete message:
+
+```shell
+postgres=> INSERT INTO ... # Long query.
+ERROR: Unable to execute query. Max query size limit (2048 characters) exceeded.
+For long queries, execute it using `tsh db` commands.
+
+postgres=>
+```
+
+### Implementation
+
+#### Web UI
+
+All the new interactions will be done on the Web UI. First, the database
+resource returned by the API will include new fields:
+
+```go
+// Defined at lib/web/ui/server.go
+type Database struct {
+  // omitted
+
+  // DatabaseRoles is the list of allowed database roles that the user can
+  // select.
+  DatabaseRoles []string `json:"database_roles"`
```

Comment:
> Addition of database_roles on a database resource level may lead to confusion.

This isn't being added to the Database resource, but it is added to the database's Web UI representation (`lib/web/ui/` package). For example, Applications behave similarly to present the user with the [available AWSRoleARNs](https://github.com/gravitational/teleport/blob/91a9f3a6b70385eb3c8031e899508cc4645e81d8/lib/web/ui/app.go#L52-L53).

About how the values will be filled, the initial idea was to fill them with the `db_roles` from the user Teleport roles.

Building the `pg_roles` as options for the users would be a good improvement; however, as it works today, it is only valuable if the user role has wildcards. Given this, we want to stick with a more straightforward implementation and add this as future work. What do you think?

> If such a list was available, it would be used to disable/hide the unavailable roles.

That's how the Web UI will work:
- If the field is populated, we only display a select field with the available options.
- If it is not populated or includes a wildcard option (`*`), the UI can present a `SelectCreateable` including the available options. This field permits the user to type a new option.

The RFD lacks some of those details, but I'll add them. Thanks for bringing this up.